sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md

```diff
@@ -0,0 +1,731 @@
+is a constant expression, as determined by the template parameter **__N** in a
+given specialization.
+
+Each simd class object is mapped to a consecutive block of general register
+files (GRF).
+
+```cpp
+namespace cm {
+namespace gen {
+template <typename __Ty, int __N> class simd {
+public:
+  using value_type = simd<__Ty, __N, 0>;
+  using element_type = __Ty;
+  using vector_type = __vector_type_t<__Ty, __N>;
+  using type = simd<__Ty, __N>;
+  static constexpr int length = __N;
+
+  // Constructors.
+  constexpr simd();
+  constexpr simd(const simd &other);
+  constexpr simd(simd &&other);
+  constexpr simd(const vector_type &__Val);
+  constexpr simd(std::initializer_list<__Ty> __Ilist) noexcept;
+  constexpr simd(__Ty __Val, __Ty __Step = __Ty()) noexcept;
+
+  // Assignment operators.
+  constexpr type &operator=(const type &) &;
+  constexpr type &operator=(type &&) &;
+
+  // Subscript operators.
+  __Ty operator[](int __i);
+  __Ty operator[](int __i) const;
+
+  // Unary operators.
+  type &operator++();
+  type operator++(int);
+  type &operator--();
+  type operator--(int);
+
+  // Binary operators.
+  auto operator +(const type &__RHS) const;
+  auto operator -(const type &__RHS) const;
+  auto operator *(const type &__RHS) const;
+  auto operator /(const type &__RHS) const;
+  type operator &(const type &__RHS) const;
+  type operator |(const type &__RHS) const;
+  type operator ^(const type &__RHS) const;
+
+  // Compound assignment.
+  type &operator +=(const type &__RHS);
+  type &operator -=(const type &__RHS);
+  type &operator *=(const type &__RHS);
+  type &operator /=(const type &__RHS);
+  type &operator &=(const type &__RHS);
+  type &operator |=(const type &__RHS);
+  type &operator ^=(const type &__RHS);
+
+  // Compare operators.
+  simd<uint16_t, __N> operator >(const type &__RHS) const;
+  simd<uint16_t, __N> operator >=(const type &__RHS) const;
+  simd<uint16_t, __N> operator <(const type &__RHS) const;
+  simd<uint16_t, __N> operator <=(const type &__RHS) const;
+  simd<uint16_t, __N> operator ==(const type &__RHS) const;
+  simd<uint16_t, __N> operator !=(const type &__RHS) const;
+
+  // Select operations.
+  template <int __Size, int __Stride>
+  simd<__Ty, __Size> select(uint16_t __Offset = 0) &&;
+  template <int __Size, int __Stride>
+  simd_view<type, region1d_t<__Ty, __Size, __Stride>>
+  select(uint16_t __i = 0) &;
+
+  // Replicate operations.
+  template <int Rep> simd<__Ty, Rep * __N> replicate();
+  template <int Rep, int W> simd<__Ty, Rep * W> replicate(uint16_t __i);
+  template <int Rep, int VS, int W>
+  simd<__Ty, Rep * W> replicate(uint16_t __i);
+  template <int Rep, int VS, int W, int HS>
+  simd<__Ty, Rep * W> replicate(uint16_t __i);
+
+  // Format operations.
+  template <typename __EltTy> auto format() &;
+  template <typename __EltTy, int __Height, int __Width> auto format() &;
+
+  // Merge operations.
+  void merge(const value_type &__Val, const __mask_type_t<__N> &__Mask);
```

Comment:
`__mask_type_t` isn't defined.

Reply:
Right., because this is not the complete  source. The documentation for individual API elements will be built from source eventually, so there should be no such problems in future.

Reply:
But it doesn't make sense to document an API which takes a type which isn't documented how it can be created. It's fine if the internals are not documented if they aren't needed for any of the documented APIs. But if the creating of the type isn't documented there is no documented way to use that API. You should either leave out that API for now or document at least the creation of that type.

Reply:
> > `-fsycl-esimd` option option added to tweak FE behavior slightly compared to normal SYCL. In the future it is expected that no such distinction will be necessary, and the need in extra -fsycl-esimd option will go away.
> 
> I didn't see this discussed in the extension documentation. I think you should add some discussion of it there, and I'd recommend changing it to `-fsycl-explicit-simd` or just `-fsycl-simd`.

`-fsycl-explicit-simd` sounds better, as it accords with the feature name, will use that then

Reply:
> But it doesn't make sense to document an API which takes a type which isn't documented how it can be created. 

@rolandschulz , I added missing definitions